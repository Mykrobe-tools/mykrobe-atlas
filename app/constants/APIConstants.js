/* @flow */

export const BASE_URL: string =
  process.env.NODE_ENV === 'production'
    ? (process.env.API_URL: any)
    : 'https://api.atlas-dev.makeandship.com';

// export const BASE_URL = 'https://localhost:3001';
// export const BASE_URL = 'http://13.69.243.89:8000';

export const USER_COOKIE_NAME = 'api.atlas.makeandship.com';
