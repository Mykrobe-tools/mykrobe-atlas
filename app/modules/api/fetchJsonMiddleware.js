/* @flow */

import 'isomorphic-fetch';

import type { JSendType } from '../../types/JSendType';

// TODO: make token access more generic
import {
  getAuthToken,
  signOut,
  AUTH_SIGNOUT,
  AUTH_SIGNOUT_SUCCESS,
} from '../auth/auth';

import {
  showNotification,
  NotificationCategories,
} from '../notifications/notifications';

import { fetchToCurl } from './fetchToCurl';

export const FETCH_JSON = '@@fetch-json-middleware/FETCH_JSON';

export class FetchJsonError extends Error {
  constructor(status, statusText, response) {
    super();
    this.name = 'FetchJsonError';
    this.status = status;
    this.statusText = statusText;
    this.response = response;
    this.message = `${status} - ${statusText}`;
  }
}

const isFetchJson = action => {
  return action.hasOwnProperty && action.hasOwnProperty(FETCH_JSON);
};

const normalizeTypes = types =>
  types.map(type => {
    if (typeof type === 'string') {
      return {
        type,
      };
    }
    return type;
  });

export const fetchJsonMiddleware = store => next => action => {
  // Do not process actions without [FETCH_JSON] property
  if (!isFetchJson(action)) {
    return next(action);
  }

  const fetchJsonParameters = action[FETCH_JSON];

  const { url, options = {}, types, debug = false } = fetchJsonParameters;

  const [REQUEST, SUCCESS, FAILURE] = normalizeTypes(types);

  next(REQUEST);

  const state = store.getState();
  const token = getAuthToken(state);

  if (!options.headers) {
    options.headers = {};
  }
  options.headers = {
    ...options.headers,
    Accept: 'application/json',
  };
  if (token) {
    options.headers = {
      ...options.headers,
      Authorization: `Bearer ${token}`,
    };
  }

  if (debug) {
    console.log(`fetch ${url} options`, options);
    const curl = fetchToCurl(url, options);
    console.log(curl);
  }

  return (async () => {
    try {
      const response = await fetch(url, options);

      if (!response.ok) {
        if (token && response.status === 401) {
          await store.dispatch(signOut());
        }
        throw new FetchJsonError(
          response.status,
          response.statusText,
          response
        );
      }

      let jsend: JSendType;

      try {
        jsend = await response.json();
      } catch (error) {
        throw new FetchJsonError(response.status, error.message, response);
      }

      const { status, message, data } = jsend;

      if (!status) {
        const json = JSON.stringify(jsend, null, 2);
        throw new FetchJsonError(
          response.status,
          `Invalid JSend response - ${json}`,
          response
        );
      }

      if (status !== 'success') {
        throw new FetchJsonError(response.status, message || data, response);
      }

      if (debug) {
        console.log('data', JSON.stringify(data, null, 2));
      }

      next({ ...SUCCESS, payload: data });
      return data;
    } catch (error) {
      next(FAILURE);
      const content = error.message || error.statusText;
      await store.dispatch(
        showNotification({
          category: NotificationCategories.ERROR,
          content,
        })
      );
      throw error;
    }
  })();
};
