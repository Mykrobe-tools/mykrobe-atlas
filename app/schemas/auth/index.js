/* @flow */

export { default as loginSchema } from './login';
export { default as resetPasswordSchema } from './resetPassword';
export { default as signupSchema } from './signup';
export { default as forgotPasswordSchema } from './forgotPassword';
