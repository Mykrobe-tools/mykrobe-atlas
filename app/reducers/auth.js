/* @flow */

import * as types from '../constants/ActionTypes';
import type { AuthType } from '../types/AuthTypes';

// isLoading = loading from keychain
// isFetching = fetching from API

const initialState: AuthType = {
  isLoading: true,
  isFetching: false,
  isAuthenticated: false,
};

function shapeState(state: AuthType): AuthType {
  const isAuthenticated: boolean = !!(state.user && state.user.token);
  const shapedState: AuthType = {
    ...state,
    isAuthenticated,
  };
  return shapedState;
}

/**
 * Create the raw state and then shape
 * @param  {AuthType=initialState} state
 * @param  {Object={}} action
 */

export default function auth(
  state: AuthType = initialState,
  action: Object = {}
) {
  const rawState = authRawState(state, action);
  const newState = shapeState(rawState);
  return newState;
}

function authRawState(state: AuthType, action: Object) {
  switch (action.type) {
    case types.AUTH_INITIALISE:
      return {
        ...state,
        isLoading: true,
      };
    case types.AUTH_INITIALISE_SUCCESS:
      return {
        ...state,
        isLoading: false,
        user: action.user,
      };
    case types.AUTH_SIGNIN:
      return {
        ...state,
        isFetching: true,
        failureReason: undefined,
      };
    case types.AUTH_SIGNIN_SUCCESS:
      return {
        ...state,
        isFetching: false,
        user: action.user,
      };
    case types.AUTH_SIGNIN_FAIL:
      return {
        ...state,
        isFetching: false,
        user: undefined,
        failureReason: action.statusText,
      };
    case types.AUTH_FORGOT_PASSWORD:
      return {
        ...state,
        isFetching: true,
        failureReason: undefined,
      };
    case types.AUTH_FORGOT_PASSWORD_SUCCESS:
      return {
        ...state,
        isFetching: false,
      };
    case types.AUTH_FORGOT_PASSWORD_FAIL:
      return {
        ...state,
        isFetching: false,
        failureReason: action.statusText,
      };
    case types.AUTH_RESET_PASSWORD:
      return {
        ...state,
        isFetching: true,
        failureReason: undefined,
      };
    case types.AUTH_RESET_PASSWORD_SUCCESS:
      return {
        ...state,
        isFetching: false,
      };
    case types.AUTH_RESET_PASSWORD_FAIL:
      return {
        ...state,
        isFetching: false,
        failureReason: action.statusText,
      };
    case types.AUTH_SIGNOUT:
      return {
        ...state,
      };
    case types.AUTH_SIGNOUT_SUCCESS:
      return {
        ...initialState,
        isLoading: false,
      };
    case types.AUTH_UPDATE_FAILURE_REASON:
      return {
        ...state,
        failureReason: action.failureReason,
      };
    case types.AUTH_DELETE_FAILURE_REASON:
      return {
        ...state,
        failureReason: undefined,
      };
    case types.AUTH_REQUEST_USER:
      return {
        ...state,
        isFetching: true,
        failureReason: undefined,
      };
    case types.AUTH_REQUEST_USER_SUCCESS:
      const user = {
        ...state.user,
        ...action.user,
      };
      return {
        ...state,
        user,
        isFetching: false,
      };
    case types.AUTH_REQUEST_USER_FAIL:
      return {
        ...state,
        isFetching: false,
        failureReason: action.statusText,
      };
    default:
      return state;
  }
}
