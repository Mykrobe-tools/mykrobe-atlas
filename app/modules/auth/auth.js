/* @flow */

// TODO: separate account creation and admin
// from sign in / out and authentication

import { createSelector } from 'reselect';
import { push } from 'react-router-redux';

import { fetchJson } from '../api';
import * as CredentialsHelpers from '../../helpers/CredentialsHelpers';
import { showNotification, NotificationCategories } from '../notifications';
import { BASE_URL } from '../../constants/APIConstants.js';
import type {
  AuthResetPasswordType,
  AuthVerificationType,
} from '../../types/AuthTypes';
import type { UserType } from '../../types/UserTypes';

export const typePrefix = 'auth/auth/';

export const AUTH_INITIALISE = `${typePrefix}AUTH_INITIALISE`;
export const AUTH_INITIALISE_SUCCESS = `${typePrefix}AUTH_INITIALISE_SUCCESS`;

export const AUTH_SIGNUP = `${typePrefix}AUTH_SIGNUP`;
export const AUTH_SIGNUP_SUCCESS = `${typePrefix}AUTH_SIGNUP_SUCCESS`;
export const AUTH_SIGNUP_FAIL = `${typePrefix}AUTH_SIGNUP_FAIL`;

export const AUTH_SIGNIN = `${typePrefix}AUTH_SIGNIN`;
export const AUTH_SIGNIN_SUCCESS = `${typePrefix}AUTH_SIGNIN_SUCCESS`;
export const AUTH_SIGNIN_FAIL = `${typePrefix}AUTH_SIGNIN_FAIL`;

export const AUTH_FORGOT_PASSWORD = `${typePrefix}AUTH_FORGOT_PASSWORD`;
export const AUTH_FORGOT_PASSWORD_SUCCESS = `${typePrefix}AUTH_FORGOT_PASSWORD_SUCCESS`;
export const AUTH_FORGOT_PASSWORD_FAIL = `${typePrefix}AUTH_FORGOT_PASSWORD_FAIL`;

export const AUTH_RESET_PASSWORD = `${typePrefix}AUTH_RESET_PASSWORD`;
export const AUTH_RESET_PASSWORD_SUCCESS = `${typePrefix}AUTH_RESET_PASSWORD_SUCCESS`;
export const AUTH_RESET_PASSWORD_FAIL = `${typePrefix}AUTH_RESET_PASSWORD_FAIL`;

export const AUTH_SIGNOUT = `${typePrefix}AUTH_SIGNOUT`;
export const AUTH_SIGNOUT_SUCCESS = `${typePrefix}AUTH_SIGNOUT_SUCCESS`;

export const AUTH_REQUEST_USER = `${typePrefix}AUTH_REQUEST_USER`;
export const AUTH_REQUEST_USER_SUCCESS = `${typePrefix}AUTH_REQUEST_USER_SUCCESS`;
export const AUTH_REQUEST_USER_FAIL = `${typePrefix}AUTH_REQUEST_USER_FAIL`;

export const AUTH_UPDATE_USER = `${typePrefix}AUTH_UPDATE_USER`;
export const AUTH_UPDATE_USER_SUCCESS = `${typePrefix}AUTH_UPDATE_USER_SUCCESS`;
export const AUTH_UPDATE_USER_FAIL = `${typePrefix}AUTH_UPDATE_USER_FAIL`;

export const AUTH_DELETE_USER = `${typePrefix}AUTH_DELETE_USER`;
export const AUTH_DELETE_USER_SUCCESS = `${typePrefix}AUTH_DELETE_USER_SUCCESS`;
export const AUTH_DELETE_USER_FAIL = `${typePrefix}AUTH_DELETE_USER_FAIL`;

export const AUTH_UPDATE_FAILURE_REASON = `${typePrefix}AUTH_UPDATE_FAILURE_REASON`;
export const AUTH_DELETE_FAILURE_REASON = `${typePrefix}AUTH_DELETE_FAILURE_REASON`;

export const AUTH_VERIFY = `${typePrefix}AUTH_VERIFY`;
export const AUTH_VERIFY_SUCCESS = `${typePrefix}AUTH_VERIFY_SUCCESS`;
export const AUTH_VERIFY_FAIL = `${typePrefix}AUTH_VERIFY_FAIL`;

// Selectors

export const getState = state => (state.auth ? state.auth.auth : undefined);

export const getIsLoading = createSelector(getState, auth => auth.isLoading);
export const getIsFetching = createSelector(getState, auth => auth.isFetching);
export const getIsAuthenticated = createSelector(
  getState,
  auth => !!(auth && auth.user && auth.user.token)
);
export const getFailureReason = createSelector(
  getState,
  auth => auth.failureReason
);
export const getUser = createSelector(getState, auth => auth.user);
export const getAuthToken = createSelector(getState, auth => {
  return auth && auth.user ? auth.user.token : undefined;
});

console.error(
  'TODO: replace access to auth values with selectors, especially isAuthenticated'
);

// Reducer

const initialState: AuthType = {
  isLoading: true,
  isFetching: false,
};

export default function reducer(state: AuthType, action: Object) {
  switch (action.type) {
    case AUTH_INITIALISE:
      return {
        ...state,
        isLoading: true,
      };
    case AUTH_INITIALISE_SUCCESS:
      return {
        ...state,
        isLoading: false,
        user: action.user,
      };
    case AUTH_SIGNIN:
      return {
        ...state,
        isFetching: true,
        failureReason: undefined,
      };
    case AUTH_SIGNIN_SUCCESS:
      return {
        ...state,
        isFetching: false,
        user: action.user,
      };
    case AUTH_SIGNIN_FAIL:
      return {
        ...state,
        isFetching: false,
        failureReason: action.statusText,
      };
    case AUTH_FORGOT_PASSWORD:
      return {
        ...state,
        isFetching: true,
        failureReason: undefined,
      };
    case AUTH_FORGOT_PASSWORD_SUCCESS:
      return {
        ...state,
        isFetching: false,
      };
    case AUTH_FORGOT_PASSWORD_FAIL:
      return {
        ...state,
        isFetching: false,
        failureReason: action.statusText,
      };
    case AUTH_RESET_PASSWORD:
      return {
        ...state,
        isFetching: true,
        failureReason: undefined,
      };
    case AUTH_RESET_PASSWORD_SUCCESS:
      return {
        ...state,
        isFetching: false,
      };
    case AUTH_RESET_PASSWORD_FAIL:
      return {
        ...state,
        isFetching: false,
        failureReason: action.statusText,
      };
    case AUTH_SIGNOUT:
      return {
        ...state,
      };
    case AUTH_SIGNOUT_SUCCESS:
      return {
        ...initialState,
        isLoading: false,
      };
    case AUTH_UPDATE_FAILURE_REASON:
      return {
        ...state,
        failureReason: action.failureReason,
      };
    case AUTH_DELETE_FAILURE_REASON:
      return {
        ...state,
        failureReason: undefined,
      };
    case AUTH_REQUEST_USER:
      return {
        ...state,
        isFetching: true,
        failureReason: undefined,
      };
    case AUTH_REQUEST_USER_SUCCESS: {
      const user = {
        ...state.user,
        ...action.user,
      };
      return {
        ...state,
        user,
        isFetching: false,
      };
    }
    case AUTH_REQUEST_USER_FAIL:
      return {
        ...state,
        isFetching: false,
        failureReason: action.statusText,
      };
    default:
      return state;
  }
}

// Side effects

//
// Auth token
//

export function loadAuth() {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_INITIALISE,
    });
    const user = CredentialsHelpers.loadUser();
    dispatch({
      type: AUTH_INITIALISE_SUCCESS,
      user,
    });
    return Promise.resolve(user);
  };
}

//
// Sign out
//

export function signOut() {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_SIGNOUT,
    });
    CredentialsHelpers.deleteUser();
    dispatch({
      type: AUTH_SIGNOUT_SUCCESS,
    });
    dispatch(
      showNotification({
        category: NotificationCategories.SUCCESS,
        content: 'You are now logged out',
      })
    );
    return Promise.resolve();
  };
}

//
// Sign in
//

export function signIn(user: UserType) {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_SIGNIN,
    });
    return dispatch(
      fetchJson(`${BASE_URL}/auth/login`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(user),
      })
    )
      .then((data: UserType) => {
        CredentialsHelpers.saveUser(data);
        dispatch({
          type: AUTH_SIGNIN_SUCCESS,
          user: data,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.SUCCESS,
            content: 'You are now logged in',
          })
        );
        return Promise.resolve(data);
      })
      .catch(error => {
        const { statusText } = error;
        dispatch({
          type: AUTH_SIGNIN_FAIL,
          statusText,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.ERROR,
            content: 'Sorry, your email/password was incorrect',
          })
        );
        return Promise.reject(error);
      });
  };
}

//
// Sign up
//

export function signUp(user: UserType) {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_SIGNUP,
    });
    return dispatch(
      fetchJson(`${BASE_URL}/users`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(user),
      })
    )
      .then(data => {
        dispatch({
          type: AUTH_SIGNUP_SUCCESS,
          user: data,
        });
        dispatch(push('/auth/success'));
        return Promise.resolve(data);
      })
      .catch(error => {
        const { statusText } = error;
        dispatch({
          type: AUTH_SIGNUP_FAIL,
          statusText,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.ERROR,
            content: statusText,
          })
        );
        return Promise.reject(error);
      });
  };
}

//
// Forgot
//

export function forgotPassword(user: UserType) {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_FORGOT_PASSWORD,
    });
    return dispatch(
      fetchJson(`${BASE_URL}/auth/forgot`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(user),
      })
    )
      .then(data => {
        dispatch({
          type: AUTH_FORGOT_PASSWORD_SUCCESS,
        });
        dispatch(push('/auth/forgotsuccess'));
        return Promise.resolve(data);
      })
      .catch(error => {
        const { statusText } = error;
        dispatch({
          type: AUTH_FORGOT_PASSWORD_FAIL,
          statusText,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.ERROR,
            content: statusText,
          })
        );
        return Promise.reject(error);
      });
  };
}

//
// reset password
//

export function resetPassword(reset: AuthResetPasswordType) {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_RESET_PASSWORD,
    });
    return dispatch(
      fetchJson(`${BASE_URL}/auth/reset`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(reset),
      })
    )
      .then(data => {
        dispatch({
          type: AUTH_RESET_PASSWORD_SUCCESS,
        });
        dispatch(push('/auth/resetsuccess'));
        return Promise.resolve(data);
      })
      .catch(error => {
        const { statusText } = error;
        dispatch({
          type: AUTH_RESET_PASSWORD_FAIL,
          statusText,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.ERROR,
            content: statusText,
          })
        );
        return Promise.reject(error);
      });
  };
}

//
// Failure reason
//

export function updateFailureReason(failureReason: string) {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_UPDATE_FAILURE_REASON,
      failureReason,
    });
  };
}

export function deleteFailureReason() {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_DELETE_FAILURE_REASON,
    });
  };
}

//
// current user
//

export function fetchCurrentUser() {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_REQUEST_USER,
    });
    return dispatch(fetchJson(`${BASE_URL}/user`))
      .then(data => {
        dispatch({
          type: AUTH_REQUEST_USER_SUCCESS,
          user: data,
        });
        return Promise.resolve(data);
      })
      .catch(error => {
        const { statusText } = error;
        dispatch({
          type: AUTH_REQUEST_USER_FAIL,
          statusText,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.ERROR,
            content: statusText,
          })
        );
        return Promise.reject(error);
      });
  };
}

export function updateCurrentUser(user: UserType) {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_UPDATE_USER,
    });
    return dispatch(
      fetchJson(`${BASE_URL}/user`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(user),
      })
    )
      .then((data: UserType) => {
        dispatch({
          type: AUTH_UPDATE_USER_SUCCESS,
          user: data,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.SUCCESS,
            content: 'Profile updated',
          })
        );
        return Promise.resolve(data);
      })
      .catch(error => {
        const { statusText } = error;
        dispatch({
          type: AUTH_UPDATE_USER_FAIL,
          statusText,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.ERROR,
            content: statusText,
          })
        );
        return Promise.reject(error);
      });
  };
}

export function deleteCurrentUser() {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_DELETE_USER,
    });
    return dispatch(
      fetchJson(`${BASE_URL}/user`, {
        method: 'DELETE',
      })
    )
      .then((data: UserType) => {
        dispatch({
          type: AUTH_DELETE_USER_SUCCESS,
          user: data,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.SUCCESS,
            content: 'User deleted',
          })
        );
        dispatch(signOut());
        return Promise.resolve(data);
      })
      .catch(error => {
        const { statusText } = error;
        dispatch({
          type: AUTH_DELETE_USER_FAIL,
          statusText,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.ERROR,
            content: statusText,
          })
        );
        return Promise.reject(error);
      });
  };
}

//
// Verify
//

export function verify(verify: AuthVerificationType) {
  return (dispatch: Function) => {
    dispatch({
      type: AUTH_VERIFY,
    });
    return dispatch(
      fetchJson(`${BASE_URL}/auth/verify`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(verify),
      })
    )
      .then(data => {
        dispatch({
          type: AUTH_VERIFY_SUCCESS,
          user: data,
        });
        dispatch(push('/auth/verifysuccess'));
        return Promise.resolve(data);
      })
      .catch(error => {
        const { statusText } = error;
        dispatch({
          type: AUTH_VERIFY_FAIL,
          statusText,
        });
        dispatch(
          showNotification({
            category: NotificationCategories.ERROR,
            content: statusText,
          })
        );
        dispatch(push('/auth/verifyfailure'));
        return Promise.reject(error);
      });
  };
}
