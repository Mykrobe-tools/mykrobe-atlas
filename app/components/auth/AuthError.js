/* @flow */

import * as React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { getError } from 'makeandship-js-common/src/modules/auth';

class AuthError extends React.Component<*> {
  render() {
    const { error } = this.props;
    return (
      <div>
        <h1>Authentication error</h1>
        <p>{error && error.statusText}</p>
        <Link to="/">Home</Link>
      </div>
    );
  }
}

AuthError.propTypes = {
  error: PropTypes.any,
};

const withRedux = connect(state => ({
  error: getError(state),
}));

export default withRedux(AuthError);
