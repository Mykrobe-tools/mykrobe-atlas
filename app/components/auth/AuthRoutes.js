/* @flow */

import * as React from 'react';
import { Route, Redirect, Switch } from 'react-router-dom';
import type { Match } from 'react-router-dom';
import urljoin from 'url-join';

import Login from './Login';
import Register from './Register';
import RegisterSuccess from './RegisterSuccess';
import Forgot from './Forgot';
import ForgotSuccess from './ForgotSuccess';
import AuthError from './AuthError';

import { withUserIsAuthenticatedRedirect } from 'makeandship-js-common/src/modules/auth/util';

const AuthenticatedLogin = withUserIsAuthenticatedRedirect(Login, '/');

const AuthRoutes = ({ match }: { match: Match }) => (
  <Switch>
    <Route path={`${match.url}/login`} component={AuthenticatedLogin} />
    <Route path={`${match.url}/register`} component={Register} />
    <Route path={`${match.url}/registersuccess`} component={RegisterSuccess} />
    <Route path={`${match.url}/forgot`} component={Forgot} />
    <Route path={`${match.url}/forgotsuccess`} component={ForgotSuccess} />
    <Route path={`${match.url}/error`} component={AuthError} />
    <Redirect from={`${match.url}`} exact to={urljoin(match.url, 'login')} />
  </Switch>
);

export default AuthRoutes;
