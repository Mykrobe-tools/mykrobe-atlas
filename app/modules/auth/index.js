/* @flow */

import { combineReducers } from 'redux';
import auth from './auth';

export {
  getAuth,
  getIsLoading,
  getIsFetching,
  getIsAuthenticated,
  getError,
  deleteError,
  getUser,
  getAuthToken,
  loadAuth,
  signOut,
  signIn,
  signUp,
  forgotPassword,
  resetPassword,
  requestCurrentUser,
  updateCurrentUser,
  deleteCurrentUser,
  verify,
} from './auth';

const reducer = combineReducers({
  auth,
});

export default reducer;
