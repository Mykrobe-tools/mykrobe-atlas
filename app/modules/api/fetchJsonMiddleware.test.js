/* @flow */

// provide fetch() method while running jest in Node
import 'isomorphic-fetch';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';

import { FETCH_JSON, fetchJsonMiddleware } from './fetchJsonMiddleware';
import { API_URL } from '../../constants/APIConstants.js';
import { SIGNOUT } from 'makeandship-js-common/src/modules/auth/auth';

const REQUEST = 'REQUEST';
const SUCCESS = 'SUCCESS';
const FAILURE = 'FAILURE';

const createMockStore = configureMockStore([thunk, fetchJsonMiddleware]);

const initialState = {
  auth: {
    auth: {
      token: {
        accessToken: 'MOCK_AUTH_TOKEN_FOR_TESTING',
      },
    },
  },
};

const debug = true;

describe('fetchJsonMiddleware', () => {
  const store = createMockStore(initialState);

  beforeEach(() => {
    store.clearActions();
    nock.cleanAll();
  });

  it('should handle basic GET request and resolve as promise', async () => {
    const meta = { test: true };
    const payload = { test: true };
    nock(API_URL)
      .get('/test/fetchJsonMiddleware')
      .reply(200, { status: 'success', data: payload });
    const result = await store.dispatch({
      [FETCH_JSON]: {
        url: `${API_URL}/test/fetchJsonMiddleware`,
        types: [{ type: REQUEST, meta }, SUCCESS, FAILURE],
        debug,
      },
    });
    const dispatchedActions = store.getActions();
    console.log(
      'dispatchedActions',
      JSON.stringify(dispatchedActions, null, 2)
    );
    expect(dispatchedActions[0].type).toEqual(REQUEST);
    expect(dispatchedActions[0].meta).toEqual(meta);
    expect(dispatchedActions[1].type).toEqual(SUCCESS);
    expect(dispatchedActions[1].payload).toEqual(payload);
    expect(result).toEqual(payload);
  });

  it('should reject empty response', async () => {
    nock(API_URL)
      .get('/test/fetchJsonMiddleware')
      .reply(200);
    try {
      await store.dispatch({
        [FETCH_JSON]: {
          url: `${API_URL}/test/fetchJsonMiddleware`,
          types: [REQUEST, SUCCESS, FAILURE],
          debug,
        },
      });
    } catch (error) {
      expect(error.name).toEqual('FetchJsonError');
      const dispatchedActions = store.getActions();
      console.log(
        'dispatchedActions',
        JSON.stringify(dispatchedActions, null, 2)
      );
      expect(dispatchedActions[0].type).toEqual(REQUEST);
      expect(dispatchedActions[1].type).toEqual(FAILURE);
    }
  });

  it('should reject non-jsend response', async () => {
    nock(API_URL)
      .get('/test/fetchJsonMiddleware')
      .reply(200, { data: { test: true } });
    try {
      await store.dispatch({
        [FETCH_JSON]: {
          url: `${API_URL}/test/fetchJsonMiddleware`,
          types: [REQUEST, SUCCESS, FAILURE],
          debug,
        },
      });
    } catch (error) {
      expect(error.name).toEqual('FetchJsonError');
      const dispatchedActions = store.getActions();
      console.log(
        'dispatchedActions',
        JSON.stringify(dispatchedActions, null, 2)
      );
      expect(dispatchedActions[0].type).toEqual(REQUEST);
      expect(dispatchedActions[1].type).toEqual(FAILURE);
    }
  });

  it('should sign out if unauthorized', async () => {
    nock(API_URL)
      .get('/test/fetchJsonMiddleware')
      .reply(401);
    try {
      await store.dispatch({
        [FETCH_JSON]: {
          url: `${API_URL}/test/fetchJsonMiddleware`,
          types: [REQUEST, SUCCESS, FAILURE],
          debug,
        },
      });
    } catch (error) {
      expect(error.name).toEqual('FetchJsonError');
      const dispatchedActions = store.getActions();
      console.log(
        'dispatchedActions',
        JSON.stringify(dispatchedActions, null, 2)
      );
      expect(dispatchedActions[0].type).toEqual(REQUEST);
      expect(dispatchedActions[1].type).toEqual(SIGNOUT);
      expect(dispatchedActions[2].type).toEqual(FAILURE);
    }
  });
});
