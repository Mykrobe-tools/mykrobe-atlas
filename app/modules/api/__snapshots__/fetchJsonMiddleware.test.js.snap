// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchJsonMiddleware should handle basic request 1`] = `
Array [
  Object {
    "type": "REQUEST",
  },
  Object {
    "payload": Object {
      "test": true,
    },
    "type": "SUCCESS",
  },
]
`;
